src/unix/linux.c

```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
 }
 
 
-/* This might return 0 if there was a problem getting the memory limit from
- * cgroups. This is OK because a return value of 0 signifies that the memory
- * limit is unknown.
- */
-static uint64_t uv__get_constrained_memory_fallback(void) {
-  return uv__read_uint64("/sys/fs/cgroup/memory/memory.limit_in_bytes");
-}
+static void uv__get_cgroup1_memory_limits(char* buf, uint64_t* high,
+                                          uint64_t* max) {
+  char filename[4097];
+  char* p;
+  char* p_end;
+
+  /* Find out the path of the memory controller. */
+  p = strchr(buf, ':');
+  while (p != NULL && memcmp(p, ":memory:", 8)) {
+    p = strchr(p, '\n');
+    if (p != NULL)
+      p = strchr(p, ':');
+  }
+
+  if (p != NULL) {
+    p = p + 8;
+    p_end = strchr(p, '\n');
+    if (p_end != NULL)
+      *p_end = '\0';
+
+    /* If the memory controller is mounted, check limits there. */
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/", p);
+    if (0 == access(filename, F_OK)) {
```

Comment:
This is a (admittedly mostly hypothetical) TOCTOU issue because the directory can get unmounted after access() but before open(). I'd be happier if it simply tried to read and handle the contingency if the files don't exist.